File: leetcode/1122-relative-sort-array.java
// https://leetcode.com/problems/relative-sort-array/

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

class Solution {
	public int[] relativeSortArray(int[] arr1, int[] arr2) {
		Map<Integer, Integer> arr1NumsFreq = new HashMap<>();
		for (int num: arr1) arr1NumsFreq.put(num, (arr1NumsFreq.getOrDefault(num, 0) + 1));

		int[] sortedArr1 = new int[arr1.length];
		int i = 0;
		for (int num: arr2) {
			while (arr1NumsFreq.get(num) > 0) {
				sortedArr1[i] = num;
				i += 1;
				arr1NumsFreq.put(num, arr1NumsFreq.get(num) - 1);
			}
			arr1NumsFreq.remove(num);
		}

		Set<Integer> keySet = arr1NumsFreq.keySet();
		int[] keys = new int[keySet.size()];
		int j = 0;
		for (Integer num: keySet) {
			keys[j++] = (int) num;
		}
		Arrays.sort(keys);

		for (int num: keys) {
			while (arr1NumsFreq.get(num) > 0) {
				sortedArr1[i] = num;
				i += 1;
				arr1NumsFreq.put(num, arr1NumsFreq.get(num) - 1);
			}
		}

		return sortedArr1;
	}
}

File: leetcode/1122-relative-sort-array.js
// https://leetcode.com/problems/relative-sort-array/

/**
 * @param {number[]} arr1
 * @param {number[]} arr2
 * @return {number[]}
 */
var relativeSortArray = function(arr1, arr2) {
  const arr1NumsFreq = {};
  for (const num of arr1) arr1NumsFreq[num] = (arr1NumsFreq[num] || 0) + 1;

  const sortedArr1 = [];
  for (const num of arr2) {
    while (arr1NumsFreq[num] > 0) {
      sortedArr1.push(num);
      arr1NumsFreq[num] -= 1;
    }
    delete arr1NumsFreq[num];
  }

  const keys = Object.keys(arr1NumsFreq).sort((a, b) => (a - b));
  
  for (const num of keys) {
    while (arr1NumsFreq[num] > 0) {
      sortedArr1.push(+num);
      arr1NumsFreq[num] -= 1;
    }
  }

  return sortedArr1;

};
